### PR TITLE
Feature/get multiple properties

### PR DESCRIPTION
**Pull request by Duslerke**
What:
- Created endpoint to get multiple properties
- Endpoint takes list of property references
- Custom validation that can be removed or replaced later
- api/v1/property?propertyReferences=123&propertyReferences=234
- When no properties are found returns empty array
- ATDD-ed the endpoint.

Why:
- Hackney needs to decide on unified way of validating.
- 404 generally refers to returning a single resources so I return 200 empty array when not found.

Notes:
- Paired with @PinkhamSoftware on this.
- Jira Ticket:  [APIF-43]

[APIF-43]: https://hackney.atlassian.net/browse/APIF-43

### CODE REVIEW

File: property-api/property-api.csproj [hunk 1]
```diff
@@ -11,6 +11,19 @@
...
+ <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Debug|AnyCPU'">
+   <WarningLevel>0</WarningLevel>
+ </PropertyGroup>
+
+
+ <ItemGroup>
+   <Compile Remove="V1\NewFolder\**" />
+   <Content Remove="V1\NewFolder\**" />
+   <EmbeddedResource Remove="V1\NewFolder\**" />
+   <None Remove="V1\NewFolder\**" />
```
Comment (cob16) [inline property-api/property-api.csproj]: is this change needed?

Comment (PinkhamSoftware) [reply property-api/property-api.csproj]: Um we can remove this I think.

Comment (Duslerke) [reply property-api/property-api.csproj]: Yes, the empty folder reference can be removed.

Regarding the upper limit for the number of retrievable properties, it looks like the URL size limit of ~2000 characters will set it's own upper limit. 

Due to current variable name being quite lengthy, the endpoint currently requires:
20 characters for each variable `&propertyReferences=` plus ~8 characters for the property reference. From the looks of it the endpoint would cap out at ~70 properties. 

I read a bit on it. It looks like the limit is configurable somewhat: https://stackoverflow.com/a/31250734

File: property-api.Tests/V1/Controllers/Property/PropertyControllerGetMultiplePropertiesTests.cs [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+            Assert.AreSame(propertyRef2, response.Properties[1].PropRef);
+        }
+
+        [TestCase(" ", "10")]
+        [TestCase("", "7")]
+        [TestCase("3", null)]
+        public void GivenAnInvalidListOfMultiplePropertyRefsItShouldReturnBadInput(string propertyRef, string propertyRef2)
+        {
+            //arrange
+            List<string> list = new List<string> { propertyRef, propertyRef2 };
```
Comment (LBHmsaitz) [inline property-api.Tests/V1/Controllers/Property/PropertyControllerGetMultiplePropertiesTests.cs]: Suggest renaming 'list' to something like 'propertyReferences'

File: property-api.Tests/V1/Controllers/Property/PropertyControllerGetMultiplePropertiesTests.cs [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+            //act
+            var actionResult = _classUnderTest.GetMultipleByReference(list);
+            //assert
+            actionResult.Should().BeOfType<BadRequestResult>();
+        }
+
+        //TO DO: valid
```
Comment (LBHmsaitz) [inline property-api.Tests/V1/Controllers/Property/PropertyControllerGetMultiplePropertiesTests.cs]: Is this TODO still pending for this PR?

File: property-api.Tests/V1/Controllers/Property/PropertyControllerGetMultiplePropertiesTests.cs (hunk 2 above)
Comment (Duslerke) [reply property-api.Tests/V1/Controllers/Property/PropertyControllerGetMultiplePropertiesTests.cs]: Yeah, it probably should be more descriptive. Renamed all of the "list" Lists into the "propertyReferences" for all "Get Multiple..." tests files in the project.

File: property-api.Tests/V1/Controllers/Property/PropertyControllerGetMultiplePropertiesTests.cs (hunk 3 above)
Comment (Duslerke) [reply property-api.Tests/V1/Controllers/Property/PropertyControllerGetMultiplePropertiesTests.cs]: No, this "TODO" was a reminder to implement validation. It's already implemented, I just forgot to remove this comment. So I removed it just now.

File: property-api.Tests/V1/Validation/GetMultiplePropertiesValidatorTests.cs [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    public class GetMultiplePropertiesValidatorTests
+    {
+        private IGetMultiplePropertiesValidator _classUnderTest;
+        
+        [SetUp]
+        public void Setup()
+        {
+            _classUnderTest = new GetMultiplePropertiesValidator();
+        }
+
```
Comment (LBHmsaitz) [inline property-api.Tests/V1/Validation/GetMultiplePropertiesValidatorTests.cs]: Suggest adding a test for validation passing

File: property-api/V1/Controllers/Property/PropertyController.cs [hunk 5]
```diff
@@ -45,6 +52,35 @@ public IActionResult GetByReference(string propertyReference)
...
+            if (!_getMultiplePropertiesValidator.Validate(propertyReferences))
+            {
+                return BadRequest();
+            }
+
+            var request = new GetMultiplePropertiesUseCaseRequest
+            {
+                PropertyRefs = propertyReferences
+            };
+            var usecaseResp = _getMultiplePropertiesUseCase.Execute(request);
```
Comment (LBHmsaitz) [inline property-api/V1/Controllers/Property/PropertyController.cs]: Suggest avoid abbreviating words as it can be harder to know what it really means. usecaseResp could be changed to useCaseResponse

File: property-api/V1/Validation/GetMultiplePropertiesValidator.cs [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+using System.Collections.Generic;
+using System;
+
+namespace property_api.V1.Validation
+{
+    public class GetMultiplePropertiesValidator : IGetMultiplePropertiesValidator
+    {
+        public bool Validate(IList<string> PropertyRefs)
+        {
+            if(PropertyRefs == null) { return false; }
```
Comment (LBHmsaitz) [inline property-api/V1/Validation/GetMultiplePropertiesValidator.cs]: Suggest to be consistent with the style, in some cases a space is used after 'if', in others there is no space. As far as I've seen through the rest of the code we've been using a space.

File: property-api.Tests/V1/Validation/GetMultiplePropertiesValidatorTests.cs (hunk 4 above)
Comment (Duslerke) [reply property-api.Tests/V1/Validation/GetMultiplePropertiesValidatorTests.cs]: Ok, added the test.

File: property-api/V1/UseCase/GetMultipleProperties/Boundaries/GetMultiplePropertiesUseCaseRequest.cs [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace property_api.V1.UseCase.GetMultipleProperties.Boundaries
+{
+    public class GetMultiplePropertiesUseCaseRequest
+    {
+        public IList<string> PropertyRefs { get;set;}
```
Comment (LBHmsaitz) [inline property-api/V1/UseCase/GetMultipleProperties/Boundaries/GetMultiplePropertiesUseCaseRequest.cs]: Suggest to be consistent with the style of the rest of the project, in this case the get/set bit would need a space in between

File: property-api/V1/Validation/GetMultiplePropertiesValidator.cs [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+
+            foreach (string propRef in PropertyRefs)
+            {
+
```
Comment (LBHmsaitz) [inline property-api/V1/Validation/GetMultiplePropertiesValidator.cs]: extra space not needed

File: property-api/V1/Validation/GetMultiplePropertiesValidator.cs [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+                if (string.IsNullOrEmpty(propRef) || string.IsNullOrWhiteSpace(propRef))
+                {
+                    return false;
+                }
+            }
+            return true;
+        }
+
```
Comment (LBHmsaitz) [inline property-api/V1/Validation/GetMultiplePropertiesValidator.cs]: extra line not needed

File: property-api/V1/Controllers/Property/PropertyController.cs (hunk 5 above)
Comment (Duslerke) [reply property-api/V1/Controllers/Property/PropertyController.cs]: I see. I'll make sure to remember that in the future. 
Just changed all of the abbreviations I could find across the files I worked on in the implementation and tests projects.

File: property-api/V1/Validation/GetMultiplePropertiesValidator.cs (hunk 6 above)
Comment (Duslerke) [reply property-api/V1/Validation/GetMultiplePropertiesValidator.cs]: Ok! Changed that.

File: property-api/V1/UseCase/GetMultipleProperties/Boundaries/GetMultiplePropertiesUseCaseRequest.cs (hunk 7 above)
Comment (Duslerke) [reply property-api/V1/UseCase/GetMultipleProperties/Boundaries/GetMultiplePropertiesUseCaseRequest.cs]: I see. I somehow missed that. Now it's changed in both boundaries.

File: property-api/V1/Validation/GetMultiplePropertiesValidator.cs (hunk 8 above)
Comment (Duslerke) [reply property-api/V1/Validation/GetMultiplePropertiesValidator.cs]: Line removed.

File: property-api/V1/Validation/GetMultiplePropertiesValidator.cs (hunk 9 above)
Comment (Duslerke) [reply property-api/V1/Validation/GetMultiplePropertiesValidator.cs]: Removed this line as well.

**Comment by PinkhamSoftware** [comment]
@LBHmsaitz @cob16 are you happy with this PR?

File: property-api.Tests/V1/UseCase/GetMultiplePropertiesUseCaseTests.cs [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+        public void GivenARequestContainingAListOfPropRefs_WhenIExecuteTheUseCase_ThenTheListOfPropRefsGetsPassedIntoTheGateway(string propertyReference, string propertyReference2)
+        {
+            //arrange
+            var propertyReferences = new List<string> { propertyReference, propertyReference2 };
+            var request = new GetMultiplePropertiesUseCaseRequest { PropertyReferences = propertyReferences };
+            //act
+            _classUnderTest.Execute(request);
+            //assert
+            _mockGetMultiplePropertiesGateway.Verify(g => g.GetMultiplePropertiesByPropertyListOfReferences(It.Is<List<string>>(arg => arg[0] == propertyReferences[0] && arg[1] == propertyReferences[1])), Times.Once);
+
```
Comment (LBHmsaitz) [inline property-api.Tests/V1/UseCase/GetMultiplePropertiesUseCaseTests.cs]: Wild empty line

File: property-api/V1/Gateways/PropertyGateway.cs [hunk 11]
```diff
@@ -32,5 +34,17 @@ public IList<Property> GetPropertyChildren(string propertyReference)
             var listChildren = children.Select(c => _factory.FromUHProperty(c)).ToList();
             return listChildren;
         }
+
+        public List<Property> GetMultiplePropertiesByPropertyListOfReferences(IList<string> propertyReferences)
+        {
+            if(propertyReferences == null)
```
Comment (LBHmsaitz) [inline property-api/V1/Gateways/PropertyGateway.cs]: inconsistency with the spacing after the if keyword

**Comment by PinkhamSoftware** [comment]
Looks good to me 👍

File: property-api.Tests/V1/Controllers/Property/PropertyControllerGetMultiplePropertiesTests.cs [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+        {
+            //arrange
+            List<string> propertyReferences = new List<string> { propertyReference, propertyReference2 };
+            //act
+            var actionResult = _classUnderTest.GetMultipleByReference(propertyReferences);
+            //assert
+            actionResult.Should().BeOfType<BadRequestResult>();
+        }
+    }
+}
```
Comment (LBHMGeorgieva) [inline property-api.Tests/V1/Controllers/Property/PropertyControllerGetMultiplePropertiesTests.cs]: There is no test that compares the json returned by the controller vs the expected to be returned json

File: property-api/property-api.csproj (hunk 1 above)
Comment (cob16) [reply property-api/property-api.csproj]: if you are sure that limit applies here if you are going to rely on an outside technical limitation like that please add it as a comment onto the controller.

File: property-api.Tests/V1/Controllers/Property/PropertyControllerGetMultiplePropertiesTests.cs (hunk 12 above)
Comment (Duslerke) [reply property-api.Tests/V1/Controllers/Property/PropertyControllerGetMultiplePropertiesTests.cs]: That's a good point. I added the test now. I put it inside the Acceptance test as it didn't fit in well in the controller tests I had.

File: property-api/property-api.csproj (hunk 1 above)
Comment (Duslerke) [reply property-api/property-api.csproj]: Just tested it. 

It looks like Microsoft Edge caps out at 2083 chars, while others (Firefox, Chrome, Opera, Brave) cap out at 8185. Which means, they should return 291 properties max. If you go over the limit, you get HTTP 414.

Is this too much? Should we limit the number?
